src/darwin/Framework/CHIP/MTRDeviceController.mm

```diff
@@ -654,6 +656,68 @@ - (void)setPairingDelegate:(id<MTRDevicePairingDelegate>)delegate queue:(dispatc
     });
 }
 
+- (void)setNocChainIssuer:(id<MTRNOCChainIssuer>)nocChainIssuer
+{
+    VerifyOrReturn([self checkIsRunning]);
+
+    dispatch_sync(_chipWorkQueue, ^{
+        VerifyOrReturn([self checkIsRunning]);
+
+        self->_operationalCredentialsDelegate->SetNocChainIssuer(nocChainIssuer);
+    });
+}
+
+- (NSNumber *)onNOCChainGenerationComplete:(NSData *)operationalCertificate
+                   intermediateCertificate:(NSData *)intermediateCertificate
+                           rootCertificate:(NSData *)rootCertificate
+                                       ipk:(NSData *)ipk
+                              adminSubject:(NSNumber *)adminSubject
+                                     error:(NSError * __autoreleasing *)error
+{
+    VerifyOrReturnValue([self checkIsRunning:error], [NSNumber numberWithUnsignedInt:CHIP_ERROR_INCORRECT_STATE.AsInteger()]);
+    VerifyOrReturnValue(operationalCertificate != nil, [NSNumber numberWithUnsignedInt:CHIP_ERROR_INVALID_ARGUMENT.AsInteger()]);
+    VerifyOrReturnValue(intermediateCertificate != nil, [NSNumber numberWithUnsignedInt:CHIP_ERROR_INVALID_ARGUMENT.AsInteger()]);
+    VerifyOrReturnValue(rootCertificate != nil, [NSNumber numberWithUnsignedInt:CHIP_ERROR_INVALID_ARGUMENT.AsInteger()]);
+
+    // use ipk and adminSubject from CommissioningParameters if not passed in
+    chip::Optional<chip::Controller::CommissioningParameters> commissioningParameters
+        = _cppCommissioner->GetCommissioningParameters();
+    VerifyOrReturnValue(
+        commissioningParameters.HasValue(), [NSNumber numberWithUnsignedInt:CHIP_ERROR_INCORRECT_STATE.AsInteger()]);
+
+    chip::Optional<chip::Crypto::AesCcm128KeySpan> ipkOptional;
+    uint8_t ipkValue[chip::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES];
+    chip::Crypto::AesCcm128KeySpan ipkTempSpan(ipkValue);
+    if (ipk != nil) {
+        VerifyOrReturnValue(
+            [ipk length] == sizeof(ipkValue), [NSNumber numberWithUnsignedInt:CHIP_ERROR_INCORRECT_STATE.AsInteger()]);
+        memcpy(&ipkValue[0], [ipk bytes], [ipk length]);
+        ipkOptional.SetValue(ipkTempSpan);
+    } else if (commissioningParameters.Value().GetIpk().HasValue()) {
+        ipkOptional.SetValue(commissioningParameters.Value().GetIpk().Value());
+    }
+
```

Comment:
Else... we have no ipk.  Is that OK?  What happens if we pass a no-value IPK to NOCChainGenerated?

Reply:
If no IPK is provided in the commissioning params and neither is it provided by the NOCChainIssuer as an input to onNOCChainGenerationComplete, then commissioning will fail. That is the expected result in such a case as an IPK is required